### PR TITLE
Ability to generate test operations for original values in the first object (continuation)

### PR DESCRIPTION
**Pull request by warpech**
Continuation of https://github.com/Starcounter-Jack/JSON-Patch/pull/226

Changes that I made compared to the original PR:

- was: extra boolean argument on "observe" and extra object argument on "generate" and - "compare"
- is: extra boolean argument "generate" and "compare"
- changed the word `inversible` to `invertible`. Please correct me if I'm wrong
- added way more tests for test operation generation

Pending questions:

1. https://github.com/immerjs/immer also registers inverse patches, but they do it in a separate patch, not within the same patch. Is the solution in this PR more or less useful?

Benchmark results:

#### Before this PR (bbd3002ac0ec296b8289590bdb5fd576dca17e0b, `npm run build && npm run bench-duplex`)

```julia
add operation
 Ops/sec: 3021278 ±0.33% Ran 154893 times in 0.051 seconds.
remove operation
 Ops/sec: 2467208 ±0.45% Ran 127226 times in 0.052 seconds.
replace operation
 Ops/sec: 2854277 ±0.28% Ran 146362 times in 0.051 seconds.
move operation
 Ops/sec: 271583 ±0.37% Ran 15241 times in 0.056 seconds.
copy operation
 Ops/sec: 214417 ±0.45% Ran 18773 times in 0.088 seconds.
test operation
 Ops/sec: 1316006 ±0.31% Ran 67352 times in 0.051 seconds.
===================
generate operation
 Ops/sec: 1680155 ±0.32% Ran 87099 times in 0.052 seconds.
generate operation and re-apply
 Ops/sec: 1483583 ±0.20% Ran 76006 times in 0.051 seconds.
compare operation
 Ops/sec: 427630 ±0.36% Ran 22101 times in 0.052 seconds.
compare operation same but deep objects
 Ops/sec: 122387928 ±0.31% Ran 6230468 times in 0.051 seconds.
```

#### After this PR

```julia
add operation
 Ops/sec: 3033581 ±0.26% Ran 155154 times in 0.051 seconds.
remove operation
 Ops/sec: 2513818 ±0.33% Ran 129460 times in 0.051 seconds.
replace operation
 Ops/sec: 2804656 ±0.31% Ran 144192 times in 0.051 seconds.
move operation
 Ops/sec: 271968 ±0.38% Ran 15132 times in 0.056 seconds.
copy operation
 Ops/sec: 215846 ±0.35% Ran 19086 times in 0.088 seconds.
test operation
 Ops/sec: 1311638 ±0.33% Ran 67526 times in 0.051 seconds.
==================
generate operation
 Ops/sec: 1705131 ±0.32% Ran 87963 times in 0.052 seconds.
generate operation and re-apply
 Ops/sec: 1452317 ±1.99% Ran 75998 times in 0.052 seconds.
compare operation
 Ops/sec: 422706 ±0.52% Ran 21843 times in 0.052 seconds.
compare operation same but deep objects
 Ops/sec: 115692461 ±0.71% Ran 6021898 times in 0.052 seconds.
================== new benchmarks start here: ==================
generate operation, invertible = true
 Ops/sec: 1607667 ±0.27% Ran 82412 times in 0.051 seconds.
generate operation and re-apply, invertible = true
 Ops/sec: 1401189 ±0.42% Ran 73320 times in 0.052 seconds.
compare operation, invertible = true
 Ops/sec: 237997 ±0.27% Ran 12237 times in 0.051 seconds.
compare operation same but deep objects, invertible = true
 Ops/sec: 116096872 ±0.50% Ran 5978518 times in 0.051 seconds.
````

### CODE REVIEW

**Comment by warpech** [comment]
cc @pytlesk4

File: src/duplex.ts [hunk 1]
```diff
@@ -185,19 +186,22 @@ function _generate(mirror, obj, patches, path) {
       var newVal = obj[key];
 
       if (typeof oldVal == "object" && oldVal != null && typeof newVal == "object" && newVal != null) {
-        _generate(oldVal, newVal, patches, path + "/" + escapePathComponent(key));
+        _generate(oldVal, newVal, patches, path + "/" + escapePathComponent(key), invertible);
       }
       else {
         if (oldVal !== newVal) {
           changed = true;
+          if (invertible) patches.push({ op: "test", path: path + "/" + escapePathComponent(key), value: _deepClone(oldVal) });
```
Comment (tomalec) [inline src/duplex.ts]: Could you add `{}` for `if`s. In my opinion it would make the code more readable.

File: test/spec/duplexSpec.js [hunk 2]
```diff
@@ -1430,45 +1605,91 @@ describe('duplex', function() {
...
           op: 'replace',
           path: '/arr/deeperArray',
           value: obj
         }
       ]);
-            
+
     });
-    it('should return an add for a property that does not exist in the first obj', function() {
+    it('should return an add for a property that does not exist in the first obj, without a test operation (invertible = true)', function() {
```
Comment (tomalec) [inline test/spec/duplexSpec.js]: I think that we should keep all the tests for `invertible = false` as well.

**Comment by warpech** [comment]
@tomalec please re-review. 

I have achieved performing the same test using `invertible` set as `true` and then `false` by creating a new helper functions: `variantIt`, `insertIf` and usage of spread operator. I think this is quite neat, similar to NUnit fixtures. I wonder if this is readable, though. Please let me know.

**Comment by tomalec** [comment]
> I wonder if this is readable, though. Please let me know.

At least to me it is :)

**Comment by pytlesk4** [comment]
@warpech thanks for taking this over the finish line!

**Comment by tomalec** [comment]
@pytlesk4 once again thanks for starting a great PR and sorry for being picky :)

**Comment by warpech** [comment]
Released today

**Comment by pytlesk4** [comment]
@warpech can this get published to NPM?

**Comment by warpech** [comment]
Done, thanks for pinging me!

**Pull request merged**